lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/evaluator/EvaluatorShim.java

```diff
@@ -0,0 +1,215 @@
+  }
+
+  public void run() {
+    LOG.log(Level.FINEST, "Entering EvaluatorShim.run().");
+    this.onStart();
+  }
+
+  public void stop() {
+    this.onStop();
+  }
+
+  @Override
+  public void onNext(final RemoteMessage<EvaluatorShimProtocol.EvaluatorShimControlProto> remoteMessage) {
+    final EvaluatorShimProtocol.EvaluatorShimCommand command = remoteMessage.getMessage().getCommand();
+    if (command == EvaluatorShimProtocol.EvaluatorShimCommand.EVALUATOR_LAUNCH) {
+      LOG.log(Level.INFO, "Received a command to launch the Evaluator.");
+      EvaluatorShim.this.onEvaluatorLaunch(remoteMessage.getMessage().getEvaluatorLaunchCommand(),
+          remoteMessage.getMessage().getEvaluatorConfigString());
+    } else if (command == EvaluatorShimProtocol.EvaluatorShimCommand.TERMINATE) {
+      LOG.log(Level.INFO, "Received a command to terminate.");
+      EvaluatorShim.this.onStop();
+    } else {
+      LOG.log(Level.WARNING, "An unknown command was received by the EvaluatorShim: {0}.",
+          command);
+      throw new IllegalArgumentException("An unknown command was received by the EvaluatorShim.");
+    }
+  }
+
+  private void onStart() {
+    LOG.log(Level.FINEST, "Entering EvaluatorShim.onStart().");
+
+    LOG.log(Level.INFO, "Reporting back to the driver with Shim Status = {0}",
+        EvaluatorShimProtocol.EvaluatorShimStatus.ONLINE);
+    this.evaluatorShimStatusChannel.onNext(
+        EvaluatorShimProtocol.EvaluatorShimStatusProto
+            .newBuilder()
+            .setRemoteIdentifier(this.remoteManager.getMyIdentifier())
+            .setContainerId(this.containerId)
+            .setStatus(EvaluatorShimProtocol.EvaluatorShimStatus.ONLINE)
+            .build());
+  }
+
+  private void onStop() {
+    LOG.log(Level.FINER, "Entering EvaluatorShim.onStop().");
+
+    if (this.evaluatorProcess != null) {
+      try {
+        LOG.log(Level.INFO, "Waiting for the evaluator process to exit.");
+        this.evaluatorProcess.waitFor();
+
+        LOG.log(Level.INFO, "Destroying the Evaluator Process.");
+        this.evaluatorProcess.destroy();
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    LOG.log(Level.INFO, "Shutting down the evaluatorLaunchExecutorService.");
+    this.evaluatorLaunchExecutorService.shutdown();
+
+    try {
+      LOG.log(Level.INFO, "Closing EvaluatorShim Control channel.");
+      this.evaluatorShimCommandChannel.close();
+    } catch (Exception e) {
+      LOG.log(Level.SEVERE, "An unexpected exception occurred while attempting to close the EvaluatorShim " +
+          "control channel.");
+      throw new RuntimeException(e);
+    }
+
+    try {
+      LOG.log(Level.INFO, "Closing the Remote Manager.");
+      this.remoteManager.close();
+    } catch (Exception e) {
+      LOG.log(Level.SEVERE, "Failed to close the RemoteManager with the following exception: {0}.", e);
```

Comment:
If you don't want to throw a new RuntimeException you should comment explicitly that it's intended and why.

Reply:
I just missed it. Will later revisit to see if we really need/want to throw exceptions in close methods. Because an exception in on part will prevent other resources from being closed.